### PR TITLE
Unit Testing

### PR DESCRIPTION
**Pull request by welzie**
* Unit Testing section (ready for review)
* TDD section (ready for review)
* Exercises (ready for review)
* Studio (ready for review)

### CODE REVIEW

File: src/chapters/unit-testing/exercises.rst [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+      shuttle!'`` for the empty string, decimal values, and non-numerical
+      inputs.
+
+#. Once the code returns the correct output for all of the suggested input
+   values, continue to *test* it with your own entries. Make sure your fixes
+   all work.
+
+Good! You are now ready to *automate* the testing process.
+
+Automatic Testing
```
Comment (welzie) [inline src/chapters/unit-testing/exercises.rst]: This is also going to hurt. In my opinion this section is also too long and more like a book section than an exercise. I would remove this as well. More about this in my overall comment.

File: src/chapters/unit-testing/exercises.rst [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+   .. note::
+
+      We do NOT need to check every possible value that is less than 5. Testing a single
+      example is sufficient to check that part of the function.
+
+#. Add two more ``it`` clauses inside ``describe`` to test what happens when
+   ``checkFive`` is passed a value greater than 5 or equal to 5.
+
+
+Try One on Your Own
```
Comment (welzie) [inline src/chapters/unit-testing/exercises.rst]: I really like this problem, however what the student needs to do needs to be explained more. We should flat out tell them to use unit tests to find the bug in the program. Tell them to write out test cases. Then write tests for each case. Then fix the bug base the test results.

File: src/chapters/unit-testing/exercises.rst [hunk 3]
```diff
@@ -0,0 +1,326 @@
+Exercises: Unit Testing
```
Comment (welzie) [inline src/chapters/unit-testing/exercises.rst]: @jimflores5, nice work overall. Having said that, I have some radical suggestions. 1) Cut out the first two sections 2) Replace them with the "try it on your own" stuff from TDD in Action. We could talk more about how to arrange this.

TDD in Action: https://github.com/LaunchCodeEducation/intro-to-professional-web-dev/blob/unit-testing/src/chapters/unit-testing/tdd-example.rst#use-tdd-to-add-these-features

File: src/chapters/unit-testing/tdd.rst [hunk 4]
```diff
@@ -0,0 +1,102 @@
+Test-Driven Development
+=======================
+We all want to write code that *works*! How do you know if your code works? You can
```
Comment (chrisbay) [inline src/chapters/unit-testing/tdd.rst]: This paragraph is somewhat redundant to the first section of the chapter. Rephrase so as to emphasize the points and remind students of them, rather than stating them for the first time.

File: src/chapters/unit-testing/tdd.rst [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+.. index:: ! TDD
+
+.. index::
+   single: TDD; test-driven development
+
+.. index::
+   single: unit testing; TDD
+
+**Test-driven development (TDD)** is a software development process where every feature is
```
Comment (chrisbay) [inline src/chapters/unit-testing/tdd.rst]: Drive home the *driven* part of TDD by using "before" or "first" to make it more explicit.

File: src/chapters/unit-testing/tdd.rst [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+Next write the unit test as if the parameter or function you just imagined already exists.
+This may seem a bit odd, but the process of thinking through how new code will be used
+helps find bugs and design flaws earlier.
+
+Now run the test! The test should fail or possibly your code will not compile because you
+have referenced something that does not exist yet.
+
+Finally you write the code that will make the new test pass. Normally this is where you
+would have started, but with TDD the new working code is the last step.
```
Comment (chrisbay) [inline src/chapters/unit-testing/tdd.rst]: Emphasize: *last*

File: src/chapters/unit-testing/tdd.rst [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+The TDD process requires a list of small features that are turned into unit tests.
+The focus on discrete aspects can lead to clean, clearly defined code.
+
+.. note::
+
+   Using the TDD process is not required when using unit tests. TDD is a process that some
+   organizations choose to use.
+
+
+The Test/Code Cycle
```
Comment (chrisbay) [inline src/chapters/unit-testing/tdd.rst]: This section could benefit from an example. It feels too abstract as-is.

File: src/chapters/unit-testing/tdd.rst [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+.. index:: ! red green refactor
+
+.. index::
+   single: TDD; red green refactor
+
+.. index::
+   single: unit testing; red green refactor
+
+While adding new features and making our code work is the main goal, we also want to write
+readable, efficient code that we are proud of. The red, green, refactor mantra
```
Comment (chrisbay) [inline src/chapters/unit-testing/tdd.rst]: Make "red, green, refactor" a vocab term.

File: src/chapters/unit-testing/tdd.rst [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+describes the process of writing tests, seeing them pass, and then making the code better.
+This process is a cycle of three steps. The red and green colors refer to test results which
+are often styled with red for failing tests and green for passing tests.
+
+1. Red -> Write a failing test.
+2. Green -> Make it pass by implementing the code.
+3. Refactor -> Make the code better.
```
Comment (chrisbay) [inline src/chapters/unit-testing/tdd.rst]: Include something like: "Since we have a test to verify our code, we can change it with confidence, knowing that any regression will be immediately identified by the test."

File: src/chapters/unit-testing/tdd.rst [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+       :alt: Graphic showing the cycle of phases from red the writing test, green making the test pass, and grey of refactoring code to be better which points back to red.
+
+       Red, green, refactor cycle.
+
+Refactoring code means to keep the same overall feature, but change how that feature
+is implemented. Examples of this are using different data structures, reducing the
+number of times needed to loop through an array, or even moving duplicate logic into
+a function so it can be reused.
+
+The refactor will likely require that you change code used in the unit test. That's ok,
```
Comment (chrisbay) [inline src/chapters/unit-testing/tdd.rst]: Per our chat, this bit can probably be removed.

File: src/chapters/unit-testing/tdd.rst [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+the refactor is also done in a TDD process. Decide what is the best way to implement the
+feature and then change the unit test to use this new idea. See the test fail, then
+implement the refactor idea. Finally see the tests pass with the refactored design.
+
+.. todo:: create our own version of this figure (I copied this from lynda site)
+
+
+Tests as Documentation
```
Comment (chrisbay) [inline src/chapters/unit-testing/tdd.rst]: I talked about this in the first section. Your description is more thorough than mine, so I recommend replacing/merging your stuff into the first section.

File: src/chapters/unit-testing/tdd-example.rst [hunk 12]
```diff
@@ -0,0 +1,560 @@
+TDD in Action
+=============
+
+Fork `this repl.it <https://repl.it/@launchcode/Transmission-processor-TDD-starter>`_
```
Comment (chrisbay) [inline src/chapters/unit-testing/tdd-example.rst]: Semantic link: "Fork <a>our starter code repl.it</a>..."

File: src/chapters/unit-testing/tdd-example.rst [hunk 13]
```diff
@@ -0,0 +1,560 @@
...
+         id: 1410,
+         rawData: 932829840830053761
+      }
+
+
+Requirements
+------------
+The features for this project have already been broken down into
+small testable units. Let's review these and then we will
+take it slow and handle these one at a time.
```
Comment (chrisbay) [inline src/chapters/unit-testing/tdd-example.rst]: these -> them

File: src/chapters/unit-testing/tdd-example.rst [hunk 14]
```diff
@@ -0,0 +1,560 @@
...
+
+#. Take in a transmission string and return an object ``{}``.
```
Comment (chrisbay) [inline src/chapters/unit-testing/tdd-example.rst]: Using `{}` could be interpreted as an empty object literal, which is not what I think you're saying here.

File: src/chapters/unit-testing/tdd-example.rst [hunk 15]
```diff
@@ -0,0 +1,560 @@
...
+#. Return ``-1`` if the transmission does NOT contain ``"::"``.
+#. Returned object should contain an ``id`` property
+
+   * The ``id`` is the part of the transmission *before* the ``"::"``
```
Comment (chrisbay) [inline src/chapters/unit-testing/tdd-example.rst]: "The value of ``id`` is the part..."

File: src/chapters/unit-testing/tdd-example.rst [hunk 16]
```diff
@@ -0,0 +1,560 @@
...
+
+#. The ``id`` property should be of type ``Number``
+#. Returned object should contain a ``rawData`` property
+
+   * The ``rawData`` is the part of the transmission *after* the ``"::"``
```
Comment (chrisbay) [inline src/chapters/unit-testing/tdd-example.rst]: "The value of ``rawData`` is the part..."

File: src/chapters/unit-testing/hello-jasmine.rst [hunk 17]
```diff
@@ -0,0 +1,301 @@
...
+#. ``index.js`` contains the Jasmine code needed to run the tests. This is the file that executes when you hit *run* in repl.it.
+
+.. warning:: Jasmine can be set up and used in many different ways. If you are looking for an answer on the Internet,---on Stack Overflow or in the Jasmine documentation---you will see widely varying usages of Jasmine that don't apply to your situation. Rely on this book as your main reference, and you'll be fine.
+
+Hello, Jasmine!
+---------------
+
+Let's build a "Hello, World!" Jasmine project, to get familiar with the basic components. Open and fork `this repl.it project <https://repl.it/@launchcode/Hello-Jasmine-Starter-Code>`_
+
+We will walk you through the steps need to get a simple Jasmine project up and running. Code along with us throughout this section.
```
Comment (jimflores5) [inline src/chapters/unit-testing/hello-jasmine.rst]: "...through the steps needed..."
or
"...through the steps you need..."

File: src/chapters/unit-testing/hello-jasmine.rst [hunk 18]
```diff
@@ -0,0 +1,301 @@
...
+
+
+``index.js``
+^^^^^^^^^^^^
+
+This is the main project file. Up until now, ``index.js`` is where you have been writing the code for a given exercise or assignment. Now that we are writing tests for our code, ``index.js`` will contain the Jasmine code to find and execute the tests. Our project-specific code will live in other files.
+
+.. sourcecode:: js
```
Comment (jimflores5) [inline src/chapters/unit-testing/hello-jasmine.rst]: The repl.it code does not match the sourcecode below.  In particular, repl.it shows ``var`` instead of ``const``.

File: src/chapters/unit-testing/hello-jasmine.rst [hunk 19]
```diff
@@ -0,0 +1,301 @@
...
+   }
+
+The ``hello`` function takes a single argument representing a person's name and returns a string greeting that person. If the function is called without an argument, the function returns ``"Hello, World!"``.
+
+To use this function outside ``hello.js`` we must export it. Add this statement at the bottom of the file.
+
+.. sourcecode:: js
+   :linenos:
+
+   module.exports = hello;
```
Comment (jimflores5) [inline src/chapters/unit-testing/hello-jasmine.rst]: Remind me to update the exercises, studio and repl.it files to match this syntax.

File: src/chapters/unit-testing/unit-testing-in-action.rst [hunk 20]
```diff
@@ -0,0 +1,218 @@
...
+When writing our ``isPalindrome`` function initially, we made two important decisions:
+
+- Case should be considered, and
+- whitespace should be considered.
+
+The definition of a palindrome differs sometimes on these two matters, so it's important to test them.
+
+Testing with input ``"abA"``` ensures that case is considered, since the lowercase version of this string, ``"aba"``, *is* a palindrome. Testing with ``"so many dynamos"`` ensures that whitespace is considered, since the version of this string with whitespace removed, ``"somanydynamos"``, *is* a palindrome.
+
+.. note:: It's important to isolate your test cases. For example, ``"So Many Dynamos"`` is a poor choice of input for a negative test, since it contains *two* characteristics that are being tested for, case and the presence of whitespace. If a test with this input failed, it would NOT be clear why it failed.
```
Comment (jimflores5) [inline src/chapters/unit-testing/unit-testing-in-action.rst]: Grammar: Ended a phrase with a preposition. 
Suggested edit: "...since it tests *two* characteristics - case and whitespace."

Also, word removal: "If a test with this input failed, it would NOT be clear why."

File: src/chapters/unit-testing/unit-testing-in-action.rst [hunk 21]
```diff
@@ -0,0 +1,218 @@
...
+
+.. pull-quote::
+
+   An edge case is a test case that provides input at the extreme edge of what the unit should be able to handle.
+
+Edge cases can look very different for different units of code. Most of the examples we provided above dealt with numerical edge cases. However, edge cases can also be non-numeric. 
+
+In the case of ``isPalindrome``, the most obvious edge case would be that of the empty string, ``""``. This is the smallest possible string that we can use when calling ``isPalindrome``. Not only is it the smallest, but it is essentially *different* from the next longest string, ``"a"``---one has characters and one doesn't. 
+
+Should the empty string be considered a palindrome? That decision is up to us, the programmer, and there is not right or wrong answer. In our case, we have decided to take a very literal definition of the term "palindrome" by considering case and whitespace. In other words, our definition says that a string is a palindrome exactly when it equals its reverse. Since the reverse of ``""`` is also ``""``, it makes sense to consider the emtpy string a palindrome.
```
Comment (jimflores5) [inline src/chapters/unit-testing/unit-testing-in-action.rst]: Typo fix: "...there is no right or wrong answer."
Passive to active fix: "In our case, we decided to..."
Word removal: "Our definition says that a string is a palindrome when it equals its reverse."

File: src/chapters/unit-testing/unit-testing-in-action.rst [hunk 22]
```diff
@@ -0,0 +1,218 @@
...
+
+You might think that another edge case is that of the longest possible palindrome. Such a palindrome would be as long as the longest possible string in JavaScript. This case is not worth considering for a couple of reasons:
+
+- The length of the longest string `can vary across different JavaScript implementations <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/length#Description>`_.
+- The most recent JavaScript specification, ES2016, states that the maximum allowed length of a string should be 2 :sup:`53` - 1 characters. This is a LOT of characters, and it is unrealistic to expect that our function will ever be given such a string.
+
+Toward a Better Testing Workflow
+--------------------------------
+
+In this case, we had a well-written function to write tests for, so it was straightforward to create tests that pass. Most situations will not be this simple. You will often find yourself writing tests for code only to discover a bug, forcing you to go back and update your code. That's okay! This is precisely what tests are for.
```
Comment (jimflores5) [inline src/chapters/unit-testing/unit-testing-in-action.rst]: Word removal and grammar fixes.
1st sentence: "In this case, we had a well-written function, so it was straightforward to create tests that passed."
3rd sentence: "Your tests will often uncover bugs in your code that need to be fixed."
Last sentence: "Do not panic! This is EXACTLY why we write tests."

File: src/chapters/unit-testing/unit-testing-in-action.rst [hunk 23]
```diff
@@ -0,0 +1,218 @@
...
+#. Write code
+#. Write tests
+#. Fix any bugs found while testing
+
+The rest of the chapter focuses on a programming technique that allows you to completly *eliminate* the third step, by reversing the order of the first two:
+
+#. Write tests
+#. Write code
+
+As you will soon learn, writing your tests *before* writing your code is a great way to enhance your programming efficiency and quality.
```
Comment (jimflores5) [inline src/chapters/unit-testing/unit-testing-in-action.rst]: Word removal: "As you will soon learn, writing your tests *before* the code is a great way to enhance your programming efficiency and quality."

File: src/chapters/unit-testing/tdd.rst [hunk 24]
```diff
@@ -0,0 +1,102 @@
...
+----------------------
+Remembering what and why your code does may not seem hard at this time, however as the
```
Comment (jimflores5) [inline src/chapters/unit-testing/tdd.rst]: Grammar fixes! "Remembering what your code does and why you structured it a certain way is easy for small programs. However, as the number of your projects increase and their size grows, the need for documentation becomes critical."

File: src/chapters/unit-testing/tdd.rst [hunk 25]
```diff
@@ -0,0 +1,102 @@
...
+number of projects increase and size of the projects grow, so does the need for documentation.
+
+Documentation can be in the form of code comments or external text documents. These can
```
Comment (jimflores5) [inline src/chapters/unit-testing/tdd.rst]: More grammar fixes.  "Documentation can take the form of code comments or external text documents. These are helpful, but they have one major drawback---they get out of date very quickly. Incorrect documentation frustrates users who try to interpret your code."

File: src/chapters/unit-testing/tdd.rst [hunk 26]
```diff
@@ -0,0 +1,102 @@
...
+be helpful, but have one major drawback which is that they can get out of date very
+quickly. Out dated, incorrect documentation is very frustrating for a user.
+
+Properly designed unit tests are runnable documentation for your project. Because unit
+tests are runnable code that declares and verifies features, they can NEVER get out of
+sync with the updated code. If feature is added or removed, the tests must be updated
```
Comment (jimflores5) [inline src/chapters/unit-testing/tdd.rst]: Typo.  "If a feature..."

File: src/chapters/unit-testing/tdd-example.rst [hunk 27]
```diff
@@ -0,0 +1,560 @@
...
+  Finished in 0.056 seconds
+
+.. admonition:: Try It!
+
+   The test data we used was missing ``<`` at the beginning. Add tests
+   to cover these cases. ``-1`` should be returned for all of these.
+
+   * ``"9701::8729740349572>0912"``
+   * ``9701::4872<97403495720912"``
+   * ``9701::487297403495720912"``
```
Comment (jimflores5) [inline src/chapters/unit-testing/tdd-example.rst]: How about adding the edge case: ``9701::<487297403<495720912>"``?

File: src/chapters/unit-testing/tdd-example.rst [hunk 28]
```diff
@@ -0,0 +1,560 @@
...
+
+
+Use TDD to Add These Features
+-----------------------------
+Use the steps demonstrated above to implement all or some of the below features.
+Take your time, you can do it!
+
+#. Trim leading and trailing whitespace from transmission.
```
Comment (jimflores5) [inline src/chapters/unit-testing/tdd-example.rst]: Missing ticks in ``transmission``.

File: src/chapters/unit-testing/tdd-example.rst [hunk 29]
```diff
@@ -0,0 +1,560 @@
...
+#. Return -1 if the id part of the transmission can not be converted to a number.
```
Comment (jimflores5) [inline src/chapters/unit-testing/tdd-example.rst]: Missing ticks in ``id`` and ``transmission``.

File: src/chapters/unit-testing/tdd-example.rst [hunk 30]
```diff
@@ -0,0 +1,560 @@
...
+#. Return -1 if more than one ``"::"`` found in transmission
```
Comment (jimflores5) [inline src/chapters/unit-testing/tdd-example.rst]: Missing ticks in ``transmission``.

File: src/chapters/unit-testing/exercises.rst [hunk 31]
```diff
@@ -0,0 +1,326 @@
...
+You probably also encountered a number of bugs in your code samples that
+required you to step through your programs to find syntax, reference, or logic
+errors. You applied ``console.log`` and other tricks to help fix the
+mistakes.
+
+The following exercises begin with a debugging review and continue with an
+introduction to automatic testing. The goal is to write code to test your code,
+rather than throwing in ``console.log`` statements to check your work.
+
+Manual Testing
```
[the remark this answers is not in the document]
Comment (jimflores5) [reply src/chapters/unit-testing/exercises.rst]: No worries, no pain. I wrote the exercises and studio before the Unit Testing chapter content was done, so I was not sure how much detail to include. We can certainly refactor the tasks.

File: src/chapters/unit-testing/exercises.rst (hunk 1 above)
Comment (jimflores5) [reply src/chapters/unit-testing/exercises.rst]: I concur. This replicates the examples from the book.  We can easily remove this part.

File: src/chapters/unit-testing/tdd-example.rst (hunk 14 above)
Comment (welzie) [reply src/chapters/unit-testing/tdd-example.rst]: The end goal is an object that has properties, but I wanted to start by having the students return an empty object. @jimflores5 how does this feel to you? I'm open to changing it if the step feels wrong.

File: src/chapters/unit-testing/why-test.rst [hunk 32]
```diff
@@ -0,0 +1,56 @@
+Why Test Your Code?
+===================
+
+Testing your code is part of the development process. Have you ever written a program and then walked away thinking, *I'm done!*, without having run it? Of course not. You are already used to testing and debugging programs as you write them. In fact, we devoted an entire chapter to :ref:`debugging <errors-and-debugging>` early in the course.
```
[the remark this answers is not in the document]
Comment (welzie) [reply src/chapters/unit-testing/why-test.rst]: I like the conversation about "walking away thinking I'm done". In my opinion it leads into the what better

File: src/chapters/unit-testing/why-test.rst [hunk 33]
```diff
@@ -0,0 +1,56 @@
...
+Your development process probably looks something like this:
+
+#. Write code
+#. Run program
+#. Notice error and investigate
+#. Repeat these steps until there are no more errors
+
+.. index:: ! unit testing
+
+But there's a better way to test your code, using *automated* tests. There are many types of automated tests. This chapter focuses on **unit testing**, which tests the smallest components (or *units*) of code. These are typically individual functions.
```
[the remark this answers is not in the document]
Comment (welzie) [reply src/chapters/unit-testing/why-test.rst]: @jimflores5 after I push take a look at what I added.

File: src/chapters/unit-testing/why-test.rst [hunk 34]
```diff
@@ -0,0 +1,56 @@
...
+
+Before we dive into the *how* of unit testing, let's discuss the *why*.
+
+Know Your Code *Really* Works
+-----------------------------
+
+Manual testing can lead you to a complete, error-free program. But unit testing does even more than that.
```
[the remark this answers is not in the document]
Comment (welzie) [reply src/chapters/unit-testing/why-test.rst]: It's not a guarantee. I agree with "can lead"

File: src/chapters/unit-testing/why-test.rst [hunk 35]
```diff
@@ -0,0 +1,56 @@
...
+
+Have you ever experienced one of this situation?
+
+.. pull-quote:: You write a program and manually test it. Thinking it is complete, you turn it in only to find that it has a bug or use case that you didn't consider.
```
[the remark this answers is not in the document]
Comment (welzie) [reply src/chapters/unit-testing/why-test.rst]: Why don't you like "turn it in"?

File: src/chapters/unit-testing/hello-jasmine.rst [hunk 36]
```diff
@@ -0,0 +1,301 @@
...
+
+We will use `Jasmine <https://jasmine.github.io/>`_, a popular JavaScript testing framework. 
+
+Using Jasmine
+-------------
+
+.. index:: 
+   single: Jasmine; installing
+
+Installing Jasmine with NPM is simple:
```
[the remark this answers is not in the document]
Comment (welzie) [reply src/chapters/unit-testing/hello-jasmine.rst]: They don't. Shortly after it states that you don't have to do this. Should we move that to be right after npm is mentioned?

File: src/chapters/unit-testing/hello-jasmine.rst [hunk 37]
```diff
@@ -0,0 +1,301 @@
...
+::
+
+   Randomized with seed 01378
+   Started
+   F
+
+   Failures:
+   1) hello should return a custom message when name is specified
+   Message:
+      AssertionError [ERR_ASSERTION]: 'Hello, Jasmine' == 'Hello, Jasmine!'
```
[the remark this answers is not in the document]
Comment (welzie) [reply src/chapters/unit-testing/hello-jasmine.rst]: I have updated this example to use output from ``assert.strictEqual``. Please review after I push

File: src/chapters/unit-testing/hello-jasmine.rst [hunk 38]
```diff
@@ -0,0 +1,301 @@
...
+   Randomized with seed 81081
+   Started
+   ..
+
+
+   2 specs, 0 failures
+   Finished in 0.025 seconds
+   Randomized with seed 81081 (jasmine --random=true --seed=81081)
+
+Nice work! You just created your first program with a full test suite. You can view `our full Hello, Jasmine! project <https://repl.it/@launchcode/Hello-Jasmine>`_ for reference.
```
[the remark this answers is not in the document]
Comment (welzie) [reply src/chapters/unit-testing/hello-jasmine.rst]: fixed in repl.it

File: src/chapters/unit-testing/unit-testing-in-action.rst [hunk 39]
```diff
@@ -0,0 +1,218 @@
...
+
+.. note:: Since we are focused on *unit* testing, in this chapter we will generally use the term "unit" to refer to the function or program under consideration.
+
+Regardless of the situation, there are two types of test cases that you should consider: positive/negative tests and edge cases.
+
+A **positive test** is a test case that gives the unit valid data, or, in the case of boolean functions, data that should result in a return value of ``true``. A **negative test** gives the unit *invalid* data, or data that should result in a return value of ``false``.
+
+An **edge case** is a test case that provides input at the extreme edge of what the unit should be able to handle. For example, if a function should work for input values between 0 and 100 then we should test not just values like 5, 50, and 75, but also 0 and 100. These values are at the literal *edge* of acceptable input values. Depending on the behavior function, we may also want to include a negative test with values -1 and 101.
```
[the remark this answers is not in the document]
Comment (welzie) [reply src/chapters/unit-testing/unit-testing-in-action.rst]: The function description doe not indicate if the test only accepts whole numbers. Do you think it needs more?

File: src/chapters/unit-testing/tdd-example.rst (hunk 14 above)
Comment (jimflores5) [reply src/chapters/unit-testing/tdd-example.rst]: I'm OK with starting with an empty object, though I would prefer that to be explicitly stated.

"Take in a transmission string and return an empty object ``{}``."
OR
""Take in a transmission string and return an object with no assigned key/value pairs ``{}``."

Comment (welzie) [reply src/chapters/unit-testing/tdd-example.rst]: So the test doesn't check if the object returned is empty, it simple checks if the type is object. Are you ok with leaving it as is?

File: src/chapters/unit-testing/why-test.rst (hunk 32 above)
[the remark this answers is not in the document]
Comment (jimflores5) [reply src/chapters/unit-testing/why-test.rst]: I understood your intent, and I agree with the idea. However, I felt that the way it was phrased chopped up the flow too much.  We can work on preserving the idea and making it fit better into the context.

File: src/chapters/unit-testing/why-test.rst (hunk 34 above)
[the remark this answers is not in the document]
Comment (jimflores5) [reply src/chapters/unit-testing/why-test.rst]: I see your point, in this case. Applied to TDD, however, "leads" would be the better option, since it reinforces the conclusion we want the students to make.

File: src/chapters/unit-testing/tdd.rst (hunk 25 above)
Comment (welzie) [reply src/chapters/unit-testing/tdd.rst]: Nice suggestion, btw this was moved to the bottom of the first page of the chapter.

File: src/chapters/unit-testing/why-test.rst (hunk 35 above)
[the remark this answers is not in the document]
Comment (jimflores5) [reply src/chapters/unit-testing/why-test.rst]: I was thinking of "turn it in" in terms of the assignments for LC101. My suggestions were to move away from this impression toward something more relevant to the workplace.

File: src/chapters/unit-testing/hello-jasmine.rst (hunk 36 above)
[the remark this answers is not in the document]
Comment (jimflores5) [reply src/chapters/unit-testing/hello-jasmine.rst]: I don't see where it states that npm is an optional install.  Regardless, I can see the students responding to "Installing with NPM" with confusion, since there is no mention of it before. I remember installing npm and Jasmine to make testing run in VSCode, but if the students do not need to install Jasmine on their machines, then avoid complicating the issue with npm.

If the students do need Jasmine, then install with something they have already used, like pip or conda, or provide a link to complete instructions (maybe in an appendix).

File: src/chapters/unit-testing/hello-jasmine.rst (hunk 37 above)
[the remark this answers is not in the document]
Comment (jimflores5) [reply src/chapters/unit-testing/hello-jasmine.rst]: Better, but if there is only one test, why does ``..F..`` appear? This seems to imply 5 specs and 1 failure.

File: src/chapters/unit-testing/unit-testing-in-action.rst (hunk 39 above)
[the remark this answers is not in the document]
Comment (jimflores5) [reply src/chapters/unit-testing/unit-testing-in-action.rst]: Good point. I was just trying to think of other edge cases.  Maybe check if a string gets passed to the function?

File: src/chapters/unit-testing/tdd-example.rst (hunk 14 above)
Comment (jimflores5) [reply src/chapters/unit-testing/tdd-example.rst]: OK.

File: src/chapters/unit-testing/tdd-example.rst [hunk 40]
```diff
@@ -0,0 +1,560 @@
...
+except *test* code at this point. There isn't anything to verify!
+
+We are going to think about and visualize
+how this feature should in code. Then we will write out in the test how
+this new code will be used. Bear with it, this exercise is actually
+quite useful.
+
+We need think of something that will satisfy this statement
+``"it("takes a string returns an object", ``.
+We have to determine what the ``it`` is. Let's define ``it`` a function named
```
[the remark this answers is not in the document]
Comment (welzie) [reply src/chapters/unit-testing/tdd-example.rst]: I lean towards keeping it as a paragraph, unless you feel it would be much easier to read as a list.

File: src/chapters/unit-testing/tdd-example.rst [hunk 41]
```diff
@@ -0,0 +1,560 @@
...
+      });
+
+   });
+
+We have an idea for a function named ``processor`` and we have imported it.
+Keep in mind this function only exists as a concept and we are writing a test
+to see if this concept makes sense.
+
+Now the real meat of the test. We are going to use ``assert.strictEqual`` to
+verify that if we pass a string to ``processor`` that an ``{}`` is returned.
```
[the remark this answers is not in the document]
Comment (welzie) [reply src/chapters/unit-testing/tdd-example.rst]: Wait until I push to verify that I have done this for all mentions of `{}` on this page.

File: src/chapters/unit-testing/tdd-example.rst (hunk 40 above)
[the remark this answers is not in the document]
Comment (jimflores5) [reply src/chapters/unit-testing/tdd-example.rst]: The list was to provide you options for three different grammar edits for the same sentence.

File: src/chapters/unit-testing/tdd-example.rst [hunk 42]
```diff
@@ -0,0 +1,560 @@
...
+   :linenos:
+
+   function process(transmission) {
+      return {};
+   }
+
+   module.exports = process;
+
+
+.. figure:: figures/processor-module-added-to-replit.png
```
[the remark this answers is not in the document]
Comment (welzie) [reply src/chapters/unit-testing/tdd-example.rst]: The figure shows extra information about where the files should go in replit.

File: src/chapters/unit-testing/exercises.rst [hunk 43]
```diff
@@ -0,0 +1,326 @@
+Exercises: Unit Testing
+========================
+
+In many of your previous coding tasks, you had to verify that your code
```
[the remark this answers is not in the document]
Comment (jimflores5) [reply src/chapters/unit-testing/exercises.rst]: I condensed the content into a single paragraph. I think it sets up the exercises and studio better.

File: src/chapters/unit-testing/exercises.rst (hunk 3 above)
Comment (jimflores5) [reply src/chapters/unit-testing/exercises.rst]: OK. We will chat again after I push up the 2nd draft.
